mibidata/mibi_image.py

```diff
@@ -185,6 +185,77 @@ def __init__(self, data, channels, **kwargs):
             setattr(self, k, v)
             self._user_defined_attributes.append(k)
 
+    def add_attr(self, **kwargs):
+        """Adds user-defined metadata key-value pairs as attributes to
+           the class instance in use. If attribute already exists
+           for the current instance, raises an error.
+
+        Args:
+            kwargs: A mapping of arguments for a user to add multiple
+                    attributes with their respecitve
+                    values.
+        Raises:
+            ValueError: Raised if attempts to set an attribute that is already
+            defined for this instance.
+
+        """
+        already_defined = []
+        for key, value in kwargs.items():
```

Comment:
Let's do check in the loop without setting the values, then raise the error if one or more attributes already exist. That way, when the user gets the error, he/she can assume the operation failed without some of the attributes having the desired value. Then, after the `if already_defined:` check passes, you can loop again and set the attribute values.